Enable local files option in OpenSCAP

jan-cerny opened the pull request:
Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=2081777

This PR adds support for OpenSCAP's `--local-files` option. This option enables users to provide local copies of remote components of SCAP source data streams. 

Unfortunately, before the introduction of this option, there existed a way to to provide local copies of remote components by simply putting them into the root directory, which now isn't possible. It turned out that this was used by foreman_scap_plugin users. Therefore, this PR is attempt to fix a regression for them.

The `--local-files` option has been introduced to OpenSCAP in upstream version 1.3.6 (January 2022). It's been introduced to upstream by https://github.com/OpenSCAP/openscap/pull/1769. I found it has never been backported to older openscap versions, therefore its present only in RPM packages openscap-1.3.6-1 and greater not present in RHEL 7, in RHEL 8 since RHEL 8.6, in RHEL 9 since RHEL 9.0.

The `--local-files` option isn't recognized by older OpenSCAP versions, they fail with an error `unrecognized option '--local-files'`. The oscap command accepts both `--local-files` and `--fetch-remote-resources` at the same time, however, it isn't clear which of the 2 options take precedence. If the `--local-files` is provided but the given directory doesn't contain the file of a name matching the data stream component name, then a warning is shown, but the scan is performed normally, only the rule `security_patches_up_to_date` is notchecked as expected.

This change is based on suggestion by @ares to add --local-files /root option. This option will be added based on the version of the openscap installed, because we need to be compatible with both old and new versions of OpenSCAP at the same time.

I have tried the patch with different client systems:
RHEL 8.7 with openscap-1.3.6-5.el8_7.x86_64
RHEL 8.4 with openscap-1.3.4-7.el8_4.x86_64
RHEL 9.1 with openscap-1.3.6-4.el9.x86_64

Question: Can we rely that the users always put the local copy of remote SCAP source data streams components to /root?

ofedoren commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,14 @@ def scan_command
-      "oscap xccdf eval #{fetch_remote_resources} #{profile} #{tailoring_subcommand} --results-arf #{results_path} #{config[@policy_id][:content_path]}"
+      # OpenSCAP 1.3.6 and newer requires the `--local-files` option to use local copies of remote SDS components
+      version = `rpm -q openscap`.split('-')[1]
+      if version >= '1.3.6' and not config[:fetch_remote_resources]
+        local_files = '--local-files /root'
+      else
+        local_files = ''
+      end
```
Even though it can't be used here, we have in Foreman itself one macro for safe version comparison: https://github.com/theforeman/foreman/blob/develop/app/services/foreman/renderer/scope/macros/base.rb#L350

You can use same logic here:
```suggestion
      local_files = if Gem::Version.new(version) >= Gem::Version.new('1.3.6') && !config[:fetch_remote_resources]
                      '--local-files /root'
                    else
                      ''
                    end
```

adamruzicka commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,14 @@ def scan_command
-      "oscap xccdf eval #{fetch_remote_resources} #{profile} #{tailoring_subcommand} --results-arf #{results_path} #{config[@policy_id][:content_path]}"
+      # OpenSCAP 1.3.6 and newer requires the `--local-files` option to use local copies of remote SDS components
+      version = `rpm -q openscap`.split('-')[1]
+      if version >= '1.3.6' and not config[:fetch_remote_resources]
```
Comparing version strings directly is a little bit wonky. Wrapping both ends in `Gem::Version.new(...)` should be a safer bet.

adamruzicka replied:
In general in ruby `&&` is usually used over `and`, which works the same way, but has lower precedence. The same goes for `!` and `not`

adamruzicka commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,14 @@ def scan_command
+      version = `rpm -q openscap`.split('-')[1]
+      if version >= '1.3.6' and not config[:fetch_remote_resources]
+        local_files = '--local-files /root'
+      else
+        local_files = ''
```
There shouldn't be a need for the else branch.

ekohl reviewed (commented):
I'm questioning if `oscap` runs on the Foreman host or on the client. This plugin runs under the Foreman process, so `rpm -q openscap` runs on host. But if `oscap` runs on the client (or even Foreman Proxy) then the check doesn't match. On the other hand, if it runs on the Foreman host (under the `foreman` user) then it wouldn't have access to `/root` at runtime.

And if this does run on the Foreman host then I think we should make it a hard RPM requirement since we don't care about EL < 8.6 since that's not supported. That simplifies the code a lot.

ofedoren commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,14 @@ def scan_command
                                end
-      "oscap xccdf eval #{fetch_remote_resources} #{profile} #{tailoring_subcommand} --results-arf #{results_path} #{config[@policy_id][:content_path]}"
+      # OpenSCAP 1.3.6 and newer requires the `--local-files` option to use local copies of remote SDS components
+      version = `rpm -q openscap`.split('-')[1]
```
Also, even though I understand that this is expected to have `openscap` installed to be able to run any scan commands this gem provides, what if `rpm -q openscap` returns `package openscap is not installed`? This will eventually raise something like `undefined method '>=' for nil:NilClass` due to `version >= '1.3.6'`. Not so helpful error in case there is no such package. I'm not familiar with this gem much, but I see there is some error handling: https://github.com/theforeman/foreman_scap_client/blob/a4245cb2317e60e328321100af85bf7a620e9ece/lib/foreman_scap_client/base_client.rb#L81-L105

ekohl replied:
I think it's a good point that you also want to suppress stderr output from the command.

ekohl replied:
Looking further at this, I'd recommend the command `rpm -q --qf '%{version}' openscap` since that only prints the version and there's no need to split it in Ruby. Then as @ofedoren suggested I'd use the `Open3.capture3` method.
```suggestion
      version, _stderr, status = Open3.capture3('rpm', '-q', '--qf', '%{version}', 'openscap')
      return false unless status.success?

```

You can also combine it into a single comparison `status.success? &&       Gem::Version.new(version) >= Gem::Version.new('1.3.6')` but I don't know if the line will be too long then.

ares commented:
> I'm questioning if oscap runs on the Foreman host or on the client. 

oscap runs on the target machine, typically not Foreman. This is the wrapper around the scanner, that's installed on every managed machine, therefore it's correct to check the rpm version. We support RHEL 6,7,8,9 on the client, so we need to detect, whether this option is already supported by the oscap tooling or not (RHEL 8.6+)

> This plugin runs...

This repository is not a foreman plugin. This lives in https://yum.theforeman.org/client/

> Question: Can we rely that the users always put the local copy of remote SCAP source data streams components to /root?

yes, currently that's how we document that (heritage of the previous oscap behavior, before --local-files existed). We wanted to keep this solution minimal and avoid introducing any configurations of custom paths etc. That would have to touch multiple repositories.

Other comments need to be resolved, the patch is in the right direction.

ekohl commented:
> oscap runs on the target machine, typically not Foreman. This is the wrapper around the scanner, that's installed on every managed machine, therefore it's correct to check the rpm version. We support RHEL 6,7,8,9 on the client, so we need to detect, whether this option is already supported by the oscap tooling or not (RHEL 8.6+)

Ah, I was confused by the `foreman_` prefix, which usually indicates a Foreman plugin but this is client tooling.

ekohl reviewed (changes requested):
Can you introduce a method `supports_local_file_option?` which abstracts away the RPM version comparison?

jan-cerny commented:
Thanks for all the reviews!

I have changed the usage of operators and I have extracted the comparison to a separate method.

ekohl commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,17 @@ def scan_command
+      supports_local_file_option ? '--local-files /root' : ''
+    end
+
+    def supports_local_file_option
```
In Ruby it's a common practice to include `?` in a method if it returns a boolean, so this is valid.
```suggestion
    def supports_local_file_option?
```

ekohl commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,17 @@ def scan_command
+    def supports_local_file_option
+      # OpenSCAP 1.3.6 and newer requires the `--local-files` option to use local copies of remote SDS components
+      version = `rpm -q openscap`.split('-')[1]
+      Gem::Version.new(version) >= Gem::Version.new('1.3.6') && !config[:fetch_remote_resources]
```
I'd keep the `config` part out since that doesn't have anything to do with support or not. That will remain even if support for version < 1.3.6 is dropped.

jan-cerny commented:
I have renamed the method to use a question mark and I have also moved the condition for fetching remote resources away from the supports_local_file_option method.

jan-cerny commented:
I have improved it to query rpm in a more robust way.

ekohl reviewed (approved):
I haven't tested this, but code wise it reads well to me. Since I gave quite a bit of direction to how it was written I'd appreciate if the other reviewers take a look as well.

ofedoren reviewed (approved):
Thanks, @jan-cerny, LGTM

ares commented:
Thanks everyone involved, merging, the release will follow.

PiRomant commented on lib/foreman_scap_client/client.rb:
```diff
@@ -36,7 +36,18 @@ def scan_command
+
+    def supports_local_file_option?
+      # OpenSCAP 1.3.6 and newer requires the `--local-files` option to use local copies of remote SDS components
+      version, _stderr, status = Open3.capture3('rpm', '-q', '--qf', '%{version}', 'openscap')
```
There is no `rpm` in Debian-based distributions.
Why don't you use 
```bash
oscap --version | grep "OpenSCAP command line tool (oscap)" |  awk '{print $NF}'
```
instead of `rpm`?
Now the whole foreman_scap_client is broken in Debian-based distributions. https://github.com/theforeman/foreman_scap_client/issues/47

adamruzicka replied:
You already filed an issue about this. Echoing the same thing in multiple places most likely will not help anything.
